File: src-tauri/src/analysis/chunk_builder.rs
/// Orchestrator: build chunks from messages using state machine classification.
///
/// The AI-buffer state machine and incremental delta logic live in their own
/// modules to keep this file focused on the high-level walk.
use crate::parsing::message_classifier::categorize_message;
use crate::parsing::metrics::calculate_metrics;
use crate::types::chunks::{EnhancedChunk, Process, SessionDetail};
use crate::types::domain::Session;
use crate::types::messages::{MessageCategory, ParsedMessage};

use super::chunk_factory::{build_compact_chunk, build_event_chunk, build_system_chunk, build_user_chunk};
use super::state_machine::ChunkBuildState;

/// Filters to main thread, classifies, and uses a state machine with AI buffer.
#[tracing::instrument(
    skip_all,
    fields(
        message_count = messages.len(),
        subagent_count = subagents.len(),
        chunk_count = tracing::field::Empty,
        elapsed_ms = tracing::field::Empty,
    )
)]
pub fn build_chunks(messages: &[ParsedMessage], subagents: &[Process]) -> Vec<EnhancedChunk> {
    let start = std::time::Instant::now();
    let mut chunks = Vec::new();
    let mut state = ChunkBuildState::default();

    let main_messages: Vec<&ParsedMessage> = messages.iter().filter(|m| !m.is_sidechain).collect();

    for msg in &main_messages {
        let category = categorize_message(msg);

        match category {
            MessageCategory::HardNoise => state.track_progress(msg),
            MessageCategory::Compact
            | MessageCategory::User
            | MessageCategory::System
            | MessageCategory::Event => {
                if let Some(ai_chunk) = state.flush_ai_chunk(subagents, messages) {
                    chunks.push(ai_chunk);
                }
                match category {
                    MessageCategory::Compact => chunks.push(build_compact_chunk(msg)),
                    MessageCategory::User => chunks.push(build_user_chunk(msg)),
                    MessageCategory::System => chunks.push(build_system_chunk(msg)),
                    MessageCategory::Event => chunks.push(build_event_chunk(msg)),
                    _ => unreachable!(),
                }
            }
            MessageCategory::Ai => state.push_ai(msg),
        }
    }

    if let Some(ai_chunk) = state.flush_ai_chunk(subagents, messages) {
        chunks.push(ai_chunk);
    }

    let span = tracing::Span::current();
    span.record("chunk_count", chunks.len());
    span.record("elapsed_ms", start.elapsed().as_millis() as u64);
    chunks
}

pub fn build_session_detail(
    session: Session,
    messages: Vec<ParsedMessage>,
    subagents: Vec<Process>,
) -> SessionDetail {
    let chunks = build_chunks(&messages, &subagents);
    let metrics = calculate_metrics(&messages);

    SessionDetail {
        session,
        messages,
        chunks,
        processes: subagents,
        metrics,
    }
}

#[cfg(test)]
#[path = "chunk_builder_tests.rs"]
mod chunk_tests;

File: internal/analysis/chunk_builder.go
package analysis

import (
	"claude-devtools/internal/domain"
	"claude-devtools/internal/parsing"
	"claude-devtools/internal/ptr"
)

// chunkBuildState buffers AI messages and flushes them into one AI chunk on the
// next non-AI message (state_machine.rs).
type chunkBuildState struct {
	aiBuffer      []domain.ParsedMessage
	progressCount uint32
	progressTexts []string
}

func (s *chunkBuildState) pushAI(m domain.ParsedMessage) {
	s.aiBuffer = append(s.aiBuffer, m)
}

func (s *chunkBuildState) trackProgress(m domain.ParsedMessage) {
	if m.MessageType != "progress" {
		return
	}
	s.progressCount++
	if m.Content.Text != nil && *m.Content.Text != "" {
		s.progressTexts = append(s.progressTexts, *m.Content.Text)
	}
}

func (s *chunkBuildState) flushAIChunk(subagents []domain.Process, all []domain.ParsedMessage) (domain.EnhancedChunk, bool) {
	if len(s.aiBuffer) == 0 {
		return domain.EnhancedChunk{}, false
	}
	var pc *uint32
	if s.progressCount > 0 {
		pc = ptr.To(s.progressCount)
	}
	var pt *[]string
	if len(s.progressTexts) > 0 {
		texts := s.progressTexts
		pt = &texts
	}
	chunk := buildAIChunkFromBuffer(s.aiBuffer, subagents, all, pc, pt)
	s.aiBuffer = nil
	s.progressCount = 0
	s.progressTexts = nil
	return chunk, true
}

// BuildChunks mirrors chunk_builder::build_chunks: classify main-thread messages
// and drive the AI-buffer state machine.
func BuildChunks(messages []domain.ParsedMessage, subagents []domain.Process) []domain.EnhancedChunk {
	chunks := []domain.EnhancedChunk{}
	state := &chunkBuildState{}

	for i := range messages {
		if messages[i].IsSidechain {
			continue
		}
		msg := messages[i]
		switch parsing.Categorize(&msg) {
		case domain.CategoryHardNoise:
			state.trackProgress(msg)
		case domain.CategoryCompact:
			chunks = appendFlush(chunks, state, subagents, messages)
			chunks = append(chunks, buildCompactChunk(&msg))
		case domain.CategoryUser:
			chunks = appendFlush(chunks, state, subagents, messages)
			chunks = append(chunks, buildUserChunk(&msg))
		case domain.CategorySystem:
			chunks = appendFlush(chunks, state, subagents, messages)
			chunks = append(chunks, buildSystemChunk(&msg))
		case domain.CategoryEvent:
			chunks = appendFlush(chunks, state, subagents, messages)
			chunks = append(chunks, buildEventChunk(&msg))
		case domain.CategoryAi:
			state.pushAI(msg)
		}
	}
	if ai, ok := state.flushAIChunk(subagents, messages); ok {
		chunks = append(chunks, ai)
	}
	return chunks
}

func appendFlush(chunks []domain.EnhancedChunk, state *chunkBuildState, subagents []domain.Process, all []domain.ParsedMessage) []domain.EnhancedChunk {
	if ai, ok := state.flushAIChunk(subagents, all); ok {
		return append(chunks, ai)
	}
	return chunks
}

// BuildSessionDetail mirrors chunk_builder::build_session_detail.
func BuildSessionDetail(session domain.Session, messages []domain.ParsedMessage, subagents []domain.Process) domain.SessionDetail {
	return domain.SessionDetail{
		Session:   session,
		Messages:  messages,
		Chunks:    BuildChunks(messages, subagents),
		Processes: subagents,
		Metrics:   parsing.CalculateMetrics(messages),
	}
}
